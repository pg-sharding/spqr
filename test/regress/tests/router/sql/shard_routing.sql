\c spqr-console

CREATE DISTRIBUTION ds1 COLUMN TYPES integer;

CREATE KEY RANGE krid1 FROM 1 ROUTE TO sh1 FOR DISTRIBUTION ds1;
CREATE KEY RANGE krid2 FROM 11 ROUTE TO sh1 FOR DISTRIBUTION ds1;
CREATE KEY RANGE krid3 FROM 21 ROUTE TO sh2 FOR DISTRIBUTION ds1;

ALTER DISTRIBUTION ds1 ATTACH RELATION xx DISTRIBUTION KEY w_id;
ALTER DISTRIBUTION ds1 ATTACH RELATION xxerr DISTRIBUTION KEY id;
ALTER DISTRIBUTION ds1 ATTACH RELATION xxtt1 DISTRIBUTION KEY w_id;

\c regress
CREATE TABLE xx (w_id int);
CREATE TABLE xxerr (i int);
CREATE TABLE xxtt1 (i int, j int, w_id int);

INSERT INTO xx (w_id) VALUES (1);
INSERT INTO xx (w_id) VALUES (10);
INSERT INTO xx (w_id) VALUES (20);
INSERT INTO xx (w_id) VALUES (21);
INSERT INTO xx (w_id) VALUES (30);


SELECT * FROM xx WHERE w_id >= 1;
SELECT * FROM xx WHERE w_id >= 20;
SELECT * FROM xx WHERE w_id >= 21;

INSERT INTO xxtt1 (w_id) VALUES(1);
INSERT INTO xxtt1 (w_id) VALUES(15);
INSERT INTO xxtt1 (w_id) VALUES(21);

INSERT INTO xxtt1 (i, w_id) VALUES(1, 1);
INSERT INTO xxtt1 (i, w_id) VALUES(15, 15);
INSERT INTO xxtt1 (i, w_id) VALUES(21, 21);

INSERT INTO xxtt1 (w_id, i) VALUES(1, 1);
INSERT INTO xxtt1 (w_id, i) VALUES(15, -12);
INSERT INTO xxtt1 (w_id, i) VALUES(21, 12);

INSERT INTO xxtt1 (i, w_id) VALUES(1, 1);
INSERT INTO xxtt1 (i, w_id) VALUES(-12, 15);
INSERT INTO xxtt1 (i, w_id) VALUES(2121221, 21);

INSERT INTO xxtt1 (i, j, w_id) VALUES(-12, 1, 1);
INSERT INTO xxtt1 (i, w_id, j) VALUES(-12, 15, 123123);
INSERT INTO xxtt1 (j, i, w_id) VALUES(2121221, -211212, 23);
INSERT INTO xxtt1 (j, i, w_id) VALUES(2121221, -211212, 21);
INSERT INTO xxtt1 (j, i, w_id) VALUES(2121221, -211212, 21);
INSERT INTO xxtt1 (j, i, w_id) VALUES(2121221, -211212, 21);
INSERT INTO xxtt1 (j, i, w_id) VALUES(2121221, -211212, 21);

SELECT * FROM xxtt1 WHERE w_id >= 1;
SELECT * FROM xxtt1 WHERE w_id >= 20;
SELECT * FROM xxtt1 WHERE w_id >= 21;
SELECT DISTINCT * FROM xxtt1 WHERE w_id=21;

-- check that aliases works
SELECT * FROM xxtt1 a WHERE a.w_id >= 1;
SELECT * FROM xxtt1 a WHERE a.w_id >= 20;
SELECT * FROM xxtt1 a WHERE a.w_id >= 21;


SELECT * FROM xxtt1 a WHERE a.w_id = 21 and j + i != 0;
SELECT * FROM xxtt1 a WHERE a.w_id = 21 and w_id <= 30 and j + i != 0;

-- check that `INSERT FROM SELECT` works
INSERT INTO xx SELECT * FROM xx a WHERE a.w_id = 20;
SELECT * FROM xx WHERE w_id >= 20;

-- check that `INSERT FROM SELECT` with constant works
INSERT INTO xx (w_id) SELECT 20;
SELECT * FROM xx WHERE w_id >= 20;
INSERT INTO xxtt1 (j, w_id) SELECT a, 20 from unnest(ARRAY[10]) a;
SELECT * FROM xxtt1 WHERE w_id = 20;

-- check that complex UPDATE works
UPDATE xxtt1 set i=a.i, j=a.j from unnest(ARRAY[(1,10)]) as a(i int, j int) where w_id=20 and xxtt1.j=a.j;
SELECT * FROM xxtt1 WHERE w_id = 20;

DROP TABLE xx;
DROP TABLE xxtt1;
DROP TABLE xxerr;

\c spqr-console
DROP DISTRIBUTION ALL CASCADE;
DROP KEY RANGE ALL;