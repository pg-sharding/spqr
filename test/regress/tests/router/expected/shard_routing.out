\c spqr-console

		SPQR router admin console
	Here you can configure your routing rules
------------------------------------------------
	You can find documentation here 
https://github.com/pg-sharding/spqr/tree/master/docs

CREATE DISTRIBUTION ds1 COLUMN TYPES integer;
    add distribution    
------------------------
 distribution id -> ds1
(1 row)

CREATE KEY RANGE krid1 FROM 1 ROUTE TO sh1 FOR DISTRIBUTION ds1;
 add key range 
---------------
 bound -> 1
(1 row)

CREATE KEY RANGE krid2 FROM 11 ROUTE TO sh1 FOR DISTRIBUTION ds1;
 add key range 
---------------
 bound -> 11
(1 row)

CREATE KEY RANGE krid3 FROM 21 ROUTE TO sh2 FOR DISTRIBUTION ds1;
 add key range 
---------------
 bound -> 21
(1 row)

ALTER DISTRIBUTION ds1 ATTACH RELATION xx DISTRIBUTION KEY w_id;
      attach table      
------------------------
 relation name   -> xx
 distribution id -> ds1
(2 rows)

ALTER DISTRIBUTION ds1 ATTACH RELATION xxerr DISTRIBUTION KEY id;
       attach table       
--------------------------
 relation name   -> xxerr
 distribution id -> ds1
(2 rows)

ALTER DISTRIBUTION ds1 ATTACH RELATION xxtt1 DISTRIBUTION KEY w_id;
       attach table       
--------------------------
 relation name   -> xxtt1
 distribution id -> ds1
(2 rows)

\c regress
CREATE TABLE xx (w_id int);
NOTICE: send query to shard(s) : sh1,sh2
CREATE TABLE xxerr (i int);
ERROR:  client processing error: error processing query 'CREATE TABLE xxerr (i int);': create table stmt ignored: no sharding rule columns found, tx status IDLE
CREATE TABLE xxtt1 (i int, j int, w_id int);
NOTICE: send query to shard(s) : sh1,sh2
INSERT INTO xx (w_id) VALUES (1);
NOTICE: send query to shard(s) : sh1
INSERT INTO xx (w_id) VALUES (10);
NOTICE: send query to shard(s) : sh1
INSERT INTO xx (w_id) VALUES (20);
NOTICE: send query to shard(s) : sh1
INSERT INTO xx (w_id) VALUES (21);
NOTICE: send query to shard(s) : sh2
INSERT INTO xx (w_id) VALUES (30);
NOTICE: send query to shard(s) : sh2
SELECT * FROM xx WHERE w_id >= 1;
NOTICE: send query to shard(s) : sh1
 w_id 
------
    1
   10
   20
(3 rows)

SELECT * FROM xx WHERE w_id >= 20;
NOTICE: send query to shard(s) : sh1
 w_id 
------
   20
(1 row)

SELECT * FROM xx WHERE w_id >= 21;
NOTICE: send query to shard(s) : sh2
 w_id 
------
   21
   30
(2 rows)

INSERT INTO xxtt1 (w_id) VALUES(1);
NOTICE: send query to shard(s) : sh1
INSERT INTO xxtt1 (w_id) VALUES(15);
NOTICE: send query to shard(s) : sh1
INSERT INTO xxtt1 (w_id) VALUES(21);
NOTICE: send query to shard(s) : sh2
INSERT INTO xxtt1 (i, w_id) VALUES(1, 1);
NOTICE: send query to shard(s) : sh1
INSERT INTO xxtt1 (i, w_id) VALUES(15, 15);
NOTICE: send query to shard(s) : sh1
INSERT INTO xxtt1 (i, w_id) VALUES(21, 21);
NOTICE: send query to shard(s) : sh2
INSERT INTO xxtt1 (w_id, i) VALUES(1, 1);
NOTICE: send query to shard(s) : sh1
INSERT INTO xxtt1 (w_id, i) VALUES(15, -12);
NOTICE: send query to shard(s) : sh1
INSERT INTO xxtt1 (w_id, i) VALUES(21, 12);
NOTICE: send query to shard(s) : sh2
INSERT INTO xxtt1 (i, w_id) VALUES(1, 1);
NOTICE: send query to shard(s) : sh1
INSERT INTO xxtt1 (i, w_id) VALUES(-12, 15);
NOTICE: send query to shard(s) : sh1
INSERT INTO xxtt1 (i, w_id) VALUES(2121221, 21);
NOTICE: send query to shard(s) : sh2
INSERT INTO xxtt1 (i, j, w_id) VALUES(-12, 1, 1);
NOTICE: send query to shard(s) : sh1
INSERT INTO xxtt1 (i, w_id, j) VALUES(-12, 15, 123123);
NOTICE: send query to shard(s) : sh1
INSERT INTO xxtt1 (j, i, w_id) VALUES(2121221, -211212, 23);
NOTICE: send query to shard(s) : sh2
INSERT INTO xxtt1 (j, i, w_id) VALUES(2121221, -211212, 21);
NOTICE: send query to shard(s) : sh2
INSERT INTO xxtt1 (j, i, w_id) VALUES(2121221, -211212, 21);
NOTICE: send query to shard(s) : sh2
INSERT INTO xxtt1 (j, i, w_id) VALUES(2121221, -211212, 21);
NOTICE: send query to shard(s) : sh2
INSERT INTO xxtt1 (j, i, w_id) VALUES(2121221, -211212, 21);
NOTICE: send query to shard(s) : sh2
SELECT * FROM xxtt1 WHERE w_id >= 1;
NOTICE: send query to shard(s) : sh1
  i  |   j    | w_id 
-----+--------+------
     |        |    1
     |        |   15
   1 |        |    1
  15 |        |   15
   1 |        |    1
 -12 |        |   15
   1 |        |    1
 -12 |        |   15
 -12 |      1 |    1
 -12 | 123123 |   15
(10 rows)

SELECT * FROM xxtt1 WHERE w_id >= 20;
NOTICE: send query to shard(s) : sh1
 i | j | w_id 
---+---+------
(0 rows)

SELECT * FROM xxtt1 WHERE w_id >= 21;
NOTICE: send query to shard(s) : sh2
    i    |    j    | w_id 
---------+---------+------
         |         |   21
      21 |         |   21
      12 |         |   21
 2121221 |         |   21
 -211212 | 2121221 |   23
 -211212 | 2121221 |   21
 -211212 | 2121221 |   21
 -211212 | 2121221 |   21
 -211212 | 2121221 |   21
(9 rows)

SELECT DISTINCT * FROM xxtt1 WHERE w_id=21;
NOTICE: send query to shard(s) : sh2
    i    |    j    | w_id 
---------+---------+------
 -211212 | 2121221 |   21
      12 |         |   21
      21 |         |   21
 2121221 |         |   21
         |         |   21
(5 rows)

-- check that aliases works
SELECT * FROM xxtt1 a WHERE a.w_id >= 1;
NOTICE: send query to shard(s) : sh1
  i  |   j    | w_id 
-----+--------+------
     |        |    1
     |        |   15
   1 |        |    1
  15 |        |   15
   1 |        |    1
 -12 |        |   15
   1 |        |    1
 -12 |        |   15
 -12 |      1 |    1
 -12 | 123123 |   15
(10 rows)

SELECT * FROM xxtt1 a WHERE a.w_id >= 20;
NOTICE: send query to shard(s) : sh1
 i | j | w_id 
---+---+------
(0 rows)

SELECT * FROM xxtt1 a WHERE a.w_id >= 21;
NOTICE: send query to shard(s) : sh2
    i    |    j    | w_id 
---------+---------+------
         |         |   21
      21 |         |   21
      12 |         |   21
 2121221 |         |   21
 -211212 | 2121221 |   23
 -211212 | 2121221 |   21
 -211212 | 2121221 |   21
 -211212 | 2121221 |   21
 -211212 | 2121221 |   21
(9 rows)

SELECT * FROM xxtt1 a WHERE a.w_id = 21 and j + i != 0;
NOTICE: send query to shard(s) : sh2
    i    |    j    | w_id 
---------+---------+------
 -211212 | 2121221 |   21
 -211212 | 2121221 |   21
 -211212 | 2121221 |   21
 -211212 | 2121221 |   21
(4 rows)

SELECT * FROM xxtt1 a WHERE a.w_id = 21 and w_id <= 30 and j + i != 0;
NOTICE: send query to shard(s) : sh2
    i    |    j    | w_id 
---------+---------+------
 -211212 | 2121221 |   21
 -211212 | 2121221 |   21
 -211212 | 2121221 |   21
 -211212 | 2121221 |   21
(4 rows)

-- check that `INSERT FROM SELECT` works
INSERT INTO xx SELECT * FROM xx a WHERE a.w_id = 20;
NOTICE: send query to shard(s) : sh1
SELECT * FROM xx WHERE w_id >= 20;
NOTICE: send query to shard(s) : sh1
 w_id 
------
   20
   20
(2 rows)

-- check that `INSERT FROM SELECT` with constant works
INSERT INTO xx (w_id) SELECT 20;
NOTICE: send query to shard(s) : sh1
SELECT * FROM xx WHERE w_id >= 20;
NOTICE: send query to shard(s) : sh1
 w_id 
------
   20
   20
   20
(3 rows)

INSERT INTO xxtt1 (j, w_id) SELECT a, 20 from unnest(ARRAY[10]) a;
NOTICE: send query to shard(s) : sh1
SELECT * FROM xxtt1 WHERE w_id = 20;
NOTICE: send query to shard(s) : sh1
 i | j  | w_id 
---+----+------
   | 10 |   20
(1 row)

-- check that complex UPDATE works
UPDATE xxtt1 set i=a.i, j=a.j from unnest(ARRAY[(1,10)]) as a(i int, j int) where w_id=20 and xxtt1.j=a.j;
NOTICE: send query to shard(s) : sh1
SELECT * FROM xxtt1 WHERE w_id = 20;
NOTICE: send query to shard(s) : sh1
 i | j  | w_id 
---+----+------
 1 | 10 |   20
(1 row)

DROP TABLE xx;
NOTICE: send query to shard(s) : sh1,sh2
DROP TABLE xxtt1;
NOTICE: send query to shard(s) : sh1,sh2
DROP TABLE xxerr;
NOTICE: send query to shard(s) : sh1,sh2
ERROR:  table "xxerr" does not exist
\c spqr-console

		SPQR router admin console
	Here you can configure your routing rules
------------------------------------------------
	You can find documentation here 
https://github.com/pg-sharding/spqr/tree/master/docs

DROP DISTRIBUTION ALL CASCADE;
   drop distribution    
------------------------
 distribution id -> ds1
(1 row)

DROP KEY RANGE ALL;
 drop key range 
----------------
(0 rows)

